datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -782,9 +786,11 @@ impl TreeNodeRewriter for CommonSubexprRewriter<'_> {
             return Ok(Transformed::new(expr, false, TreeNodeRecursion::Jump));
         }
 
-        let (up_index, expr_id) = &self.id_array[self.down_index];
+        let (up_index, expr_id, aliased) = &self.id_array[self.down_index];
         self.down_index += 1;
 
+        self.aliased |= aliased;
```

Comment:
This looks suspicuous to me. We don't need to alias if any of the ancestor nodes is an alias, do we? But the `f_down()` is called DFS way so if we calculate `self.aliased` as `self.aliased |= aliased` then we can take into account other nodes than just the ancestors in 2+ childrens...

Reply:
No, we don't need to alias any children of an Alias. Such nested aliases are only needed for CSE, other rules deal with the parent expression and its alias, not the children/nested exprs.

I don't get what you mean by nodes other than the ancestor. The rewriter is called on single exprs, not the entire plan at once.

Reply:
Let me give you an example tomorrow.

Reply:
So I was refering to expression trees like: `(((a + b) - c) as "alias1") * (a + b)`

I think currently in this case `a + b` gets extracted as a common expression and the original expression becomes:
`(((common_1 as "a + b") - c) as "alias1") * (common_1 as "a + b")`
(`common_1` is not the actual name of the exracted expression but let's use if for the sake of simplicity.)

Now, instead of the above this PR could drop the first added alias (`as  "a + b"`) becuase the first `a + b` has an alias ancestor node (`as "alias1"`), but it shouldn't drop the 2nd alias because the 2nd `a + b` doesn't have an alias ancestor.

Using a flag (`self.aliased`) caluclated in `f_down()`, to decide if we can drop an alias won't work. This is because `f_down()` gets called on all nodes in the subtree of the first child (`((a + b) - c) as "alias1"`) of `*` before it gets called on the 2nd child (`a + b`). So when `f_down()` encounters the 2nd `a + b`, `self.aliased` is true and you miss adding a necessary alias.

What I think this PR should do is maintain a stack of booleans and once `f_down()` if called push a new boolean item and when `f_up()` is called pop one. A new item should deppend on the previous item and if the current `expr` is `Alias`.

(An alternative to maintaing a stack could be to use the `..._with_payload` APIs to propagate down some kind of "has alias ancestor" state, but I haven't finished adding that API yet: https://github.com/apache/datafusion/pull/8664.

Reply:
Ok, I see what you mean.

I can get behind implementing the `f_up`, but the expression you provided doesn't even parse. If you can provide a query that shows the need for an alias stack I'd be happy to implement it.

In other words, can you think of a test case where Alias is a child and not the root of an expr?

Reply:
You can build up any kinds of plans in a test with `LogicalPlanBuilder` (including non-root aliases). Whether it ever actually happens or not in real query is a different question... But think of the current `CommonSubexprEliminate` rule, it did produce such non-root aliases so there can be other rules doing the same in certain scenarios despite the original query couldn't cointain such alias.

BTW I'm also ok with checking only the root node if we want to keep this rule simple and avoid using stacks, but in that case the check shouldn't be in `f_down()` as it's called on all nodes.

Reply:
If the stack solution is too complicated, where do you suggest adding the root node check?

Reply:
Right before you call `expr.rewrite()` so you can pass the boolean flag to `CommonSubexprRewriter`.

Reply:
Pushed the stack idea (used a counter instead), will let you know once the CI finishes.

Reply:
Yeah, it looks nice and simple!